public/components/notebooks/components/__tests__/notebook.test.tsx

```diff
@@ -63,27 +76,290 @@ describe('<Notebook /> spec', () => {
+    await waitFor(() => {
+      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
+    });
     expect(utils.container.firstChild).toMatchSnapshot();
-    utils.getByText('Add code block').click();
-    utils.getByText('Add visualization').click();
+
+    act(() => {
+      fireEvent.click(utils.getByText('Reporting actions'));
+    });
+
+    expect(utils.queryByTestId('download-notebook-pdf')).toBeInTheDocument();
+
+    act(() => {
+      fireEvent.click(utils.getByText('Reporting actions'));
+    });
+
+    await waitFor(() => {
+      expect(utils.queryByTestId('download-notebook-pdf')).toBeNull();
+    });
   });
 
-  it('renders the component', async () => {
-    const pplService = new PPLService(httpClientMock);
-    const setBreadcrumbs = jest.fn();
-    const renameNotebook = jest.fn();
-    const cloneNotebook = jest.fn();
-    const deleteNotebook = jest.fn();
-    const setToast = jest.fn();
-    const location = jest.fn();
-    const history = jest.fn() as any;
-    history.replace = jest.fn();
+  it('renders the empty component and checks code block operations', async () => {
+    httpClient.get = jest.fn(() => Promise.resolve((emptyNotebook as unknown) as HttpResponse));
+    let postFlag = 1;
+    httpClient.post = jest.fn(() => {
+      if (postFlag === 1) {
+        postFlag += 1;
+        return Promise.resolve((addCodeBlockResponse as unknown) as HttpResponse);
+      } else return Promise.resolve((runCodeBlockResponse as unknown) as HttpResponse);
+    });
+    const utils = render(
+      <Notebook
+        pplService={pplService}
+        openedNoteId="mock-id"
+        DashboardContainerByValueRenderer={jest.fn()}
+        http={httpClient}
+        parentBreadcrumb={{ href: 'parent-href', text: 'parent-text' }}
+        setBreadcrumbs={setBreadcrumbs}
+        renameNotebook={renameNotebook}
+        cloneNotebook={cloneNotebook}
+        deleteNotebook={deleteNotebook}
+        setToast={setToast}
+        location={location}
+        history={history}
+      />
+    );
+    await waitFor(() => {
+      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
+    });
+
+    act(() => {
+      utils.getByText('Add code block').click();
+    });
+
+    await waitFor(() => {
+      expect(utils.getByPlaceholderText(codePlaceholderText)).toBeInTheDocument();
+    });
+
+    act(() => {
+      utils.getByLabelText('Toggle show input').click();
+    });
+
+    await waitFor(() => {
+      expect(utils.queryByPlaceholderText(codePlaceholderText)).toBeNull();
+    });
+
+    act(() => {
+      utils.getByLabelText('Toggle show input').click();
+    });
+
+    act(() => {
+      fireEvent.input(utils.getByPlaceholderText(codePlaceholderText), {
+        target: { value: '%md \\n hello' },
+      });
+      fireEvent.click(utils.getByText('Run'));
+    });
+
+    await waitFor(() => {
+      expect(utils.queryByText('Run')).toBeNull();
+      expect(utils.getByText('hello')).toBeInTheDocument();
```

Comment:
This is a lot of asserts/different things being tested for one test. I'd like to see more atomic tests. Other tests in this file are similar.

Reply:
Split the tests further for code block checks.